File: 71.simplify-path.java
/*
 * @lc app=leetcode id=71 lang=java
 *
 * [71] Simplify Path
 */
class Solution {
    public String simplifyPath(String path) {
        if (path == null) return null;
        String[] dirs = path.split("/");
        Stack<String> stack = new Stack<>();
        for (String dir : dirs) {
            //System.out.println("[" + dir + "]");
            switch (dir) {
                case "":
                case ".":
                    break;
                case "..":
                    if (!stack.empty()) stack.pop();
                    break;
                default:
                    stack.push(dir);
            }
        }
        StringBuffer buffer = new StringBuffer();
        while (!stack.empty()) {
            buffer.insert(0, "/" + stack.pop());
        }
        return buffer.length() == 0 ? "/" : buffer.toString();
    }
}

File: go/71.simplify-path.go
import "strings"

/*
 * @lc app=leetcode id=71 lang=golang
 *
 * [71] Simplify Path
 */

// @lc code=start
func simplifyPath(path string) string {
	var stack []string
	s := strings.Split(path, "/")
	for _, ss := range s {
		switch ss {
		case "":
		case ".":
		case "..":
			if len(stack) > 0 {
				stack = stack[:len(stack)-1]
			}
		default:
			stack = append(stack, ss)
		}
	}
	res := ""
	for _, ss := range stack {
		res += "/" + ss
	}
	if res == "" {
		res = "/"
	}
	return res
}

// @lc code=end

